### PR TITLE
[Heatmap] Implement feature #13536

### PR DESCRIPTION
**Pull request by joelios**
For this feature to work properly, the [Frappe pull request](https://github.com/frappe/frappe/pull/5387)  must also be merged!

status:
 - [x] customer
 - [x] item
 - [x] company
 - [x] supplier
 - [x] Employee
- [x] Project
 - [x] Member
- [x] Student

### CODE REVIEW

**Comment by joelios** [comment]
belongs to feature #13536

File: erpnext/buying/doctype/supplier/supplier.py [hunk 1]
```diff
@@ -48,3 +48,74 @@ def on_trash(self):
...
+		timestamp = get_timestamp(date)
+		out.update({ timestamp: count })
+
+	'''purchase order'''
+	items = dict(frappe.db.sql('''select transaction_date, count(*)
+		from `tabPurchase Order` where supplier=%s
+			and transaction_date > date_sub(curdate(), interval 1 year)
+			group by transaction_date''', name))
+
+	for date, count in iteritems(items):
```
Comment (pratu16x7) [inline erpnext/buying/doctype/supplier/supplier.py]: Thanks for the contribution :) While the finer control and overriding in every Doctype is great, I'm slightly worried about code duplication; for example, this `for` functionality that takes in `items` and updates `out` could be commonified as a utility function `def update_timeline_data(items, out)` that is always called with the most recently updated `out`. There are 3-4 occurrence of these same 5 lines across 8 files, I'm sure they can be reduced :)

**Comment by pratu16x7** [comment]
Can you check the Codacy review?

File: erpnext/buying/doctype/supplier/supplier.py (hunk 1 above)
Comment (pratu16x7) [reply erpnext/buying/doctype/supplier/supplier.py]: Another immediate side effect is having to use a `for` to update `out`; there are 3-4 occurrence of these same 5 lines of the loop across 8 files.

**Comment by nabinhait** [comment]
Closing this as of now. Please reopen after fixes as per @pratu16x7 's suggestion.

File: erpnext/projects/doctype/project/project.py [hunk 2]
```diff
@@ -267,13 +267,129 @@ def update_dependencies_on_duplicated_project(self):
...
+
+	for date, count in iteritems(items):
+		timestamp = get_timestamp(date)
+		if not timestamp in out:
+			out.update({timestamp: count})
+		else :
+			out.update({timestamp: out[timestamp] + count})
+
+	'''issue'''
+	items = dict(frappe.db.sql('''select creation, count(*)
```
Comment (rmehta) [inline erpnext/projects/doctype/project/project.py]: also might be a good idea to move this into functions as this query has a standard pattern.

**Comment by rmehta** [comment]
also move this into a single function, it will shrink the volume of code by almost 80% 😉 

```
+	'''purchase receipt'''
+	items = dict(frappe.db.sql('''select posting_date, count(*)
+		from `tabPurchase Receipt` where supplier=%s
+			and posting_date > date_sub(curdate(), interval 1 year)
+			group by posting_date''', name))
+
+	for date, count in iteritems(items):
+		timestamp = get_timestamp(date)
+		if not timestamp in out:
+			out.update({timestamp: count})
+		else :
+			out.update({timestamp: out[timestamp] + count}) 
```

**Comment by bghayad** [comment]
Hello @joelios 
Hope you are doing fine.
Is it possible to continue the PR so it can be merged and used instead of keep modifying the codes manually?
Regards
Bilal

**Pull request closed without merging**